python/triton/language/semantic.py

```diff
@@ -795,96 +795,196 @@ def str_to_eviction_policy(eviction_policy):
+        else:
+            # Load by de-referencing the pointer of scalar
+            dst_ty = elt_ty
+
+        # Build IR for non-tiled load
+        if not mask:
+            return tl.tensor(builder.create_load(ptr.handle, cache, eviction, is_volatile), dst_ty)
+        else:
+            return tl.tensor(builder.create_masked_load(ptr.handle, mask.handle, other.handle if other else None,
+                                                        cache, eviction, is_volatile), dst_ty)
 
 
 def store(ptr: tl.tensor,
           val: tl.tensor,
           mask: Optional[tl.tensor],
+          boundary_check,
           cache_modifier: str,
           eviction_policy: str,
           builder: ir.builder) -> tl.tensor:
-    if not ptr.type.scalar.is_ptr():
-        raise ValueError("Pointer argument of store instruction is " + ptr.type.__repr__())
-    if not ptr.type.is_block():
-        if val.type.is_block():
-            raise ValueError("Value argument cannot be block type if pointer argument is not a block")
-        if mask and mask.type.is_block():
-            raise ValueError("Mask argument cannot be block type if pointer argument is not a block")
-    if ptr.type.is_block():
-        val = broadcast_impl_shape(val, ptr.type.get_block_shapes(), builder)
-    if mask and ptr.type.is_block():
-        mask = broadcast_impl_shape(mask, ptr.type.get_block_shapes(), builder)
-    ptr_ty = ptr.type.scalar
-    elt_ty = ptr_ty.element_ty
-    # treat bool* as tl.int8*
-    if elt_ty == tl.int1:
-        elt_ty = tl.int8
-        ptr_ty = tl.pointer_type(elt_ty, ptr_ty.address_space)
-        ptr = cast(ptr, ptr_ty, builder)
-    # attributes
+    # Cache and eviction options
     cache = str_to_cache_modifier(cache_modifier)
     eviction = str_to_eviction_policy(eviction_policy)
-    # cast to target data-type
-    val = cast(val, elt_ty, builder)
-    if not mask:
-        return tl.tensor(builder.create_store(ptr.handle, val.handle, cache, eviction), tl.void)
-    if not mask.type.scalar.is_bool():
-        raise ValueError("Mask must have boolean scalar type")
-    return tl.tensor(builder.create_masked_store(ptr.handle, val.handle, mask.handle, cache, eviction), tl.void)
+
+    if ptr.type.is_ptr() and ptr.type.element_ty.is_block():
+        # Store a tile pointer: `pointer_type<block_type<>>`
+        # Tile pointer can not have the `mask` argument
+        if mask:
+            raise ValueError("`mask` and `other` arguments cannot be specified for loading tile pointer")
+
+        # Check same shape
+        tile_shape = ptr.type.element_ty.get_block_shapes()
+        if not val.type.is_block():
+            val = broadcast_impl_shape(val, tile_shape, builder)
+        assert val.type.is_block(), "Value argument must be block type or a scalar"
+        assert tile_shape == val.type.get_block_shapes(), "Tile shape and value shape mismatch"
+
+        # TODO(Chenggang): support `tl.int1` casting for tiled store
```

Comment:
Sure.

Reply:
Fixed by casting `ptr<int1>` to `ptr<int8>` in `make_block_ptr` (but not here). Here in `load`, we still have to assert it is not a `ptr<block<int1>>` (because `make_block_ptr` must rewrite it before).

Reply:
Tests are also added with `tl.int1` cases.